src/metabase/models/action.clj

```diff
@@ -202,42 +202,72 @@
                                               ::pk? (isa? (:semantic_type field) :type/PK)})))]]
             [(:id card) parameters]))))
 
+(defn- hydrate-implicit-parameters
+  "Returns actions with generated dimplicit parameters.
+  `implicit-action-models` must contain only the models of the implicit actions."
+  [implicit-action-models actions]
+  (let [model-id->implicit-parameters (implicit-action-parameters implicit-action-models)]
+    (for [action actions]
+      (if (= (:type action) :implicit)
+        (let [saved-params    (m/index-by :id (:parameters action))
+              implicit-params (get model-id->implicit-parameters (:model_id action))
+              implicit-params (for [param implicit-params]
+                                (merge param (get saved-params (:id param))))
+              implicit-params (cond->> implicit-params
+                                (= "row/delete" (:kind action))
+                                (filter ::pk?)
+
+                                (contains? #{"row/update" "row/delete"} (:kind action))
+                                (map (fn [param] (cond-> param (::pk? param) (assoc :required true))))
+
+                                :always
+                                (map #(dissoc % ::pk? ::field-id)))]
+          (m/assoc-some action :parameters (seq implicit-params)))
```

Comment:
I probably would've written L213-225 as one continues `cond->>` pipe, but :shrug: (also done `(#{"row/update" "row/delete"} (:kind action))` instead of the explicit `contains?`)